shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Control/Iterate/SetAlgebra.hs

```diff
@@ -1,43 +1,49 @@
 {-# LANGUAGE FlexibleContexts #-}
-{-# OPTIONS_GHC -Wno-orphans #-}
-{-# OPTIONS_GHC -Wno-unused-imports #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving  #-}
+
+{-# OPTIONS_GHC -fno-warn-unused-matches #-}
+-- This option is on for only one reason, It allows us to make a list of
+-- functions, written as (\ w x y z -> Property) where each function uses
+-- a different set of the arguments. By typeing the list as
+-- [Int -> Key -> Map Key Int -> Set key -> Property], I can specify
+-- the type of each of the parameters to the anonymous functions, without
+-- repeating them for each of the over 100 items in the list.
```

Comment:
Hmmm, I feel like unused-matches is a really helpful warning, and that we can solve the `\ w x y z -> Property` problem by writing the tests in this style (where quickcheck supplies arguments with arbitrary instances until only Property is left): https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/Tripping/CBOR.hs#L161

Reply:
Every test has to have a monomorphic type, to wrap it in testProperty. To do this in the style requested, it requires 3 lines

foo :: concrete monomorphic type
foo x y = testProperty "name" ...

all_tests [ foo, 
               .... 
              ]

The current method uses the same type for all tests, the downside is that some tests
don't mention all the variables. Which can only lead to false Failures, so I think it is reasonable thing to do.